storage/parquet/datasource.go

```diff
@@ -0,0 +1,313 @@
+			break;
+		}
+	}
+
+	if it.col.MaxR() > 0 {
+		s = elements
+	}
+	return s, nil
+}
+
+func NewDataSourceBuilderFactory() physical.DataSourceBuilderFactory {
+	return physical.NewDataSourceBuilderFactory(
+		func(ctx context.Context, matCtx *physical.MaterializationContext, dbConfig map[string]interface{}, filter physical.Formula, alias string) (execution.Node, error) {
+			path, err := config.GetString(dbConfig, "path")
+			if err != nil {
+				return nil, errors.Wrap(err, "couldn't get path")
+			}
+
+			return &DataSource{
+				path:  path,
+				alias: alias,
+			}, nil
+		},
+		nil,
+		availableFilters,
+		metadata.BoundedFitsInLocalStorage,
+	)
+}
+
+// NewDataSourceBuilderFactoryFromConfig creates a data source builder factory using the configuration.
+func NewDataSourceBuilderFactoryFromConfig(dbConfig map[string]interface{}) (physical.DataSourceBuilderFactory, error) {
+	return NewDataSourceBuilderFactory(), nil
+}
+
+func (ds *DataSource) Get(ctx context.Context, variables octosql.Variables) (execution.RecordStream, error) {
+	file, err := parquet.OpenFile(ds.path)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't open file")
+	}
+	columns := file.Schema.Columns()
+	n := len(columns)
+	for _, col := range columns {
+		if col.MaxR() > 1 {
+			return nil, fmt.Errorf("column '%s' has nested repeated elements", col)
+		}
+	}
+
+	colIters := make([]*ColStrIter, n)
+	for i, col := range columns {
+		colIters[i] = NewColStrIter(file, col)
+	}
+
+	return &RecordStream{
+		file:            file,
+		isDone:          false,
+		columnIterators: colIters,
+		columns:         columns,
+		alias:           ds.alias,
+	}, nil
+}
+
+type RecordStream struct {
+	file            *parquet.File
+	columnIterators []*ColStrIter
+	columns         []parquet.Column
+	isDone          bool
+	alias           string
+}
+
+func (rs *RecordStream) Close() error {
+	err := rs.file.Close()
+	if err != nil {
+		return errors.Wrap(err, "couldn't close underlying file")
+	}
+
+	return nil
+}
+
+func (rs *RecordStream) Next(ctx context.Context) (*execution.Record, error) {
+	if rs.isDone {
+		err := rs.file.Close()
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't close underlying file")
+		}
+		return nil, execution.ErrEndOfStream
+	}
+
+	m := make(map[string]interface{})
+
+	rs.isDone = true
+
+	for i, it := range rs.columnIterators {
+		s, err := it.Next()
+		if err != nil {
+			if err == io.EOF {
+				s = nil
+			} else {
+				return nil, errors.Wrap(err, "couldn't decode parquet record")
+			}
+		} else {
+			rs.isDone = false
+		}
+		m[rs.columns[i].String()] = s
+	}
+
+	if rs.isDone == true {
+		err := rs.file.Close()
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't close underlying file")
+		}
+		return nil, execution.ErrEndOfStream
+	}
+
+	aliasedRecord := make(map[octosql.VariableName]octosql.Value)
+	for k, v := range m {
+		if str, ok := v.(string); ok {
+			parsed, err := time.Parse(time.RFC3339, str)
+			if err == nil {
```

Comment:
You should check whether `err != nil` instead, try to rewrite this way

Reply:
this code is copied from the json data source and I think with err != nil check this code will be less readable

Reply:
sure